[Snyk] Security upgrade express from 4.16.2 to 4.21.2

JetJet13 opened the pull request:
![snyk-top-banner](https://redirect.github.com/andygongea/OWASP-Benchmark/assets/818805/c518c423-16fe-447e-b67f-ad5a49b5d123)

### Snyk has created this PR to fix 1 vulnerabilities in the npm dependencies of this project.

#### Snyk changed the following file(s):

- `package.json`
- `package-lock.json`




#### Vulnerabilities that will be fixed with an upgrade:

|  | Issue | Score | 
:-------------------------:|:-------------------------|:-------------------------
![medium severity](https://res.cloudinary.com/snyk/image/upload/w_20,h_20/v1561977819/icon/m.png 'medium severity') | Regular Expression Denial of Service (ReDoS) <br/>[SNYK-JS-PATHTOREGEXP-8482416](https://snyk.io/vuln/SNYK-JS-PATHTOREGEXP-8482416) | &nbsp;&nbsp;**738**&nbsp;&nbsp; 




---

> [!IMPORTANT]
>
> - Check the changes in this PR to ensure they won't cause issues with your project.
> - Max score is 1000. Note that the real score may have changed since the PR was raised.
> - This PR was automatically created by Snyk using the credentials of a real user.

---

**Note:** _You are seeing this because you or someone else with access to this repository has authorized Snyk to open fix PRs._

For more information: <img src="https://api.segment.io/v1/pixel/track?data=eyJ3cml0ZUtleSI6InJyWmxZcEdHY2RyTHZsb0lYd0dUcVg4WkFRTnNCOUEwIiwiYW5vbnltb3VzSWQiOiIzYTBmNDdiZi01MTJlLTRlMTctYTNmNS03MzYwNzk2MzM0YWMiLCJldmVudCI6IlBSIHZpZXdlZCIsInByb3BlcnRpZXMiOnsicHJJZCI6IjNhMGY0N2JmLTUxMmUtNGUxNy1hM2Y1LTczNjA3OTYzMzRhYyJ9fQ==" width="0" height="0"/>
🧐 [View latest project report](https://app.snyk.io/org/jetjet13/project/4551e675-bccb-48f8-bccf-b17ec5d655cd?utm_source&#x3D;github&amp;utm_medium&#x3D;referral&amp;page&#x3D;fix-pr)
📜 [Customise PR templates](https://docs.snyk.io/scan-using-snyk/pull-requests/snyk-fix-pull-or-merge-requests/customize-pr-templates?utm_source=github&utm_content=fix-pr-template)
🛠 [Adjust project settings](https://app.snyk.io/org/jetjet13/project/4551e675-bccb-48f8-bccf-b17ec5d655cd?utm_source&#x3D;github&amp;utm_medium&#x3D;referral&amp;page&#x3D;fix-pr/settings)
📚 [Read about Snyk's upgrade logic](https://docs.snyk.io/scan-with-snyk/snyk-open-source/manage-vulnerabilities/upgrade-package-versions-to-fix-vulnerabilities?utm_source=github&utm_content=fix-pr-template)

---

**Learn how to fix vulnerabilities with free interactive lessons:**

🦉 [Regular Expression Denial of Service (ReDoS)](https://learn.snyk.io/lesson/redos/?loc&#x3D;fix-pr)

[//]: # 'snyk:metadata:{"customTemplate":{"variablesUsed":[],"fieldsUsed":[]},"dependencies":[{"name":"express","from":"4.16.2","to":"4.21.2"}],"env":"prod","issuesToFix":["SNYK-JS-PATHTOREGEXP-8482416"],"prId":"3a0f47bf-512e-4e17-a3f5-7360796334ac","prPublicId":"3a0f47bf-512e-4e17-a3f5-7360796334ac","packageManager":"npm","priorityScoreList":[738],"projectPublicId":"4551e675-bccb-48f8-bccf-b17ec5d655cd","projectUrl":"https://app.snyk.io/org/jetjet13/project/4551e675-bccb-48f8-bccf-b17ec5d655cd?utm_source=github&utm_medium=referral&page=fix-pr","prType":"fix","templateFieldSources":{"branchName":"default","commitMessage":"default","description":"default","title":"default"},"templateVariants":["updated-fix-title","priorityScore"],"type":"auto","upgrade":["SNYK-JS-PATHTOREGEXP-8482416"],"vulns":["SNYK-JS-PATHTOREGEXP-8482416"],"patch":[],"isBreakingChange":false,"remediationStrategy":"vuln"}'

JetJet13 commented on src/tutorials/Intro-Sets-Functions/exercises.json:
```diff
@@ -0,0 +1,30 @@
+[
```
For the exercises, we should keep a consistent format when displaying the answers, which is to display the `\square` at the end of the proof/example. Here you have squared the answer itself which is not consistent with the [Tutorial Style Guide](https://github.com/JetJet13/mathbook/wiki/Tutorial-Style-Guide#end-of-proof).

Also, the order of the exercises must be ordered by difficulty in ascending order, meaning that the easiest questions will need to be at the top and the most challenging last. I could argue that finding the cardinality is easier than finding the codomain and image of g.

rich1126 replied:
I was meaning to change that order, thanks for catching that. Squares fixed, order changed.

JetJet13 replied:
Thanks for making the changes. One thing I would like to highlight is for the Challenge question, you have the square at the end of the sentence (after the `Note`), not the actual answer.

JetJet13 commented on src/tutorials/Intro-Sets-Functions/content.json:
```diff
@@ -0,0 +1,16 @@
+    {
+        "id": "sectionBox_unfdq5xspd",
+        "contentIndex": 0,
+        "title": "Sets",
```
I think the title `Sets` is a little vague in my opinion. Does it make sense to give the section on sets a more explicit title, i.e. `Understanding Sets` or something similar ?

rich1126 replied:
I totally agree. I changed this to Understanding Sets, as suggested. That seems to be the best option.

JetJet13 commented on src/tutorials/Intro-Sets-Functions/content.json:
```diff
@@ -0,0 +1,16 @@
+    {
+        "id": "sectionBox_rqanf5bit6",
+        "contentIndex": 1,
+        "title": "Functions",
```
I think the same applies for the `Functions` section. I think a more explicit title would be best. What do you think ?

rich1126 replied:
Agree again. I changed this to "Understanding Functions" for consistency.

JetJet13 commented on src/tutorials/Intro-Sets-Functions/content.json:
```diff
@@ -0,0 +1,16 @@
+        "id": "sectionBox_rqanf5bit6",
+        "contentIndex": 1,
+        "title": "Functions",
+        "text": "<p>A function is a rule that takes elements of one set (often called <i>inputs</i>) and assigns them to elements of another set (often called <i>outputs</i>.) The first rule to always remember is that a function only has a single output for each input. </p>\n<p> Let's first look at a simple example of a function, where we take elements in our set \\(S = \\{1,2,3\\}\\) from before ...
```
I noticed that you use the term `image` in one of the paragraphs near the end of the Functions section and there doesn't seem to be any definition or minor explanation of what the image of a function is which I think might throw off the reader. I Think it would be good if we were to add some small definition for `image`.

![review-mentions-image](https://user-images.githubusercontent.com/8130149/37261453-ff44cb5a-2572-11e8-9da6-f460b7706c69.jpg)

rich1126 replied:
That is my fault. I was going back and forth between using range and image. I changed the definition of range to also include the fact that it is commonly called the image, but I will change this instance to range for clarity.

JetJet13 commented on src/tutorials/Intro-Sets-Functions/content.json:
```diff
@@ -0,0 +1,16 @@
+        "id": "sectionBox_rqanf5bit6",
+        "contentIndex": 1,
+        "title": "Functions",
+        "text": "<p>A function is a rule that takes elements of one set (often called <i>inputs</i>) and assigns them to elements of another set (often called <i>outputs</i>.) The first rule to always remember is that a function only has a single output for each input. </p>\n<p> Let's first look at a simple example of a function, where we take elements in our set \\(S = \\{1,2,3\\}\\) from before ...
```
I also found a small typo in one of your sentences,

![functions-paragraph-typo](https://user-images.githubusercontent.com/8130149/37261413-aa2d310c-2572-11e8-88cc-51b06537933f.jpg)

rich1126 replied:
Great catch. Fixed.

JetJet13 commented on src/tutorials/Intro-Sets-Functions/content.json:
```diff
@@ -0,0 +1,16 @@
+        "id": "sectionBox_rqanf5bit6",
+        "contentIndex": 1,
+        "title": "Functions",
+        "text": "<p>A function is a rule that takes elements of one set (often called <i>inputs</i>) and assigns them to elements of another set (often called <i>outputs</i>.) The first rule to always remember is that a function only has a single output for each input. </p>\n<p> Let's first look at a simple example of a function, where we take elements in our set \\(S = \\{1,2,3\\}\\) from before ...
```
Lastly, for the definition of `Range`, in braces you put `(also commonly called the range)` which seems a little redundant. Is there a reason that you wanted to include that there ?

![range-definition](https://user-images.githubusercontent.com/8130149/37261441-e53c43c8-2572-11e8-8762-8c34848e4c5f.jpg)

rich1126 replied:
Sorry; mentioned above I had been going between image and range. It now says (also commonly called image.)

rich1126 commented on src/tutorials/Intro-Sets-Functions/content.json:
```diff
@@ -0,0 +1,16 @@
+        "id": "sectionBox_unfdq5xspd",
+        "contentIndex": 0,
+        "title": "Sets",
+        "text": "<p>Most simply, a set is a collection of objects. We can think of the set of all flowers in Hawaii, or the set of whole numbers between 10 and 37. Typically, we use curly brackets (or braces) to denote a set, such as \\(\\{1,2,3\\}\\). If we are using the same set many times in a row, or talking about a set that cannot easily be written down, we can use some other symbol. Through ...
```
Good catch. This is fixed now.

rich1126 commented:
Overall, I appreciate the feedback. I added an extra paragraph to the opening of the "Understanding Sets" section to hopefully illuminate where sets fit into everything. It's brief, so if you want me to elaborate more or add some extra information, I totally can. I did the same for functions.

JetJet13 commented on src/tutorials/Intro-Sets-Functions/content.json:
```diff
@@ -2,15 +2,15 @@
-        "title": "Sets",
-        "text": "<p>Most simply, a set is a collection of objects. We can think of the set of all flowers in Hawaii, or the set of whole numbers between 10 and 37. Typically, we use curly brackets (or braces) to denote a set, such as \\(\\{1,2,3\\}\\). If we are using the same set many times in a row, or talking about a set that cannot easily be written down, we can use some other symbol. Through ...
+        "title": "Understanding Sets",
+        "text": "<p>In any field of mathematics, it is important to be able to deal with <i>objects</i> and <i>structures</i>. At the lowest level of mathematical objects and structures are sets. Most simply, a set is a collection of objects. We can think of the set of all flowers in Hawaii, or the set of whole numbers between 10 and 37. Typically, we use curly brackets (or braces) to denote a se ...
```
I found a small typo in the definition of a subset.
![subset-definition-typo](https://user-images.githubusercontent.com/8130149/37285995-0ef8e9be-25d7-11e8-9aa9-90584cc9b4e6.jpg)

JetJet13 commented on src/tutorials/Intro-Sets-Functions/content.json:
```diff
@@ -2,15 +2,15 @@
-        "title": "Functions",
-        "text": "<p>A function is a rule that takes elements of one set (often called <i>inputs</i>) and assigns them to elements of another set (often called <i>outputs</i>.) The first rule to always remember is that a function only has a single output for each input. </p>\n<p> Let's first look at a simple example of a function, where we take elements in our set \\(S = \\{1,2,3\\}\\) from before ...
+        "title": "Understanding Functions",
+        "text": "<p>Now that we have sets, we want to start to create, define, and communicate relationships and patterns <i>between</i> sets. Mathematics uses the language of a function to describe these relationships.<p>\n<p>A function is a rule that takes elements of one set (often called <i>inputs</i>) and assigns them to elements of another set (often called <i>outputs</i>.) The first rule t ...
```
Well said! 👍 
![understanding-functions-intro](https://user-images.githubusercontent.com/8130149/37286057-3eb91ca0-25d7-11e8-8b46-0a310f43f6f1.jpg)

JetJet13 replied:
Also, thanks for updating the definition of Range to include `image`.
![range-definition-update](https://user-images.githubusercontent.com/8130149/37286098-60c90198-25d7-11e8-9394-694e14e2a6ff.jpg)

JetJet13 commented:
Nice work @rich1126. I am going to merge this later today in case anyone else wants to go over it.

snyk-bot committed:
fix: package.json & package-lock.json to reduce vulnerabilities

The following vulnerabilities are fixed with an upgrade:
- https://snyk.io/vuln/SNYK-JS-PATHTOREGEXP-8482416
